source/develop/release-management/features/sla/scheduling-weight-normalization.html.md

```diff
@@ -0,0 +1,104 @@
+
+As you see in the cpu and memory example we have to either normalize the weights reported by all policy units (either manually in the code or automatically) or use some different reduction step.
+
+I was reading a bit and the three basic normalization options we can use are:
+
+- Using percentage of hardcoded maximum values for each policy unit
+- Using percentage of the maximum reported value (during the scheduling attempt) of a policy unit
+- Using rank (sorting reported weights and awarding points according to the position; my example calculation computes rank as number of hosts that are better than the ranked one)
+
+All three options have advantages and disadvantages. Hardcoded maximum is obviously a problem for future memory size values, dynamic maximum depends on which hosts make it into the weighting step and rank does not do a good job when there are huge gaps in weight values.
+
+Personally I am in favour of testing rank though as it is future proof and the factor multiplier has predictable results (and we already have the code too).
+
+#### Algorithm results comparison
+
+This is a simple case where all three algorithms report the same result. That won’t always be so.
+
+* CPU, factor 10 (hardcoded maximum 100%)
+* Memory, factor 1 (hardcoded maximum 4096 MB)
+
+
+| Host / Metric | CPU load | Occupied memory | Normalized (lower is better) | Normalized dynamic (lower is better) | Ranked (lower is better) |
+| ------------- | -------- | --------------- | ---------------------------- | ------------------------------------ | ------------------------ |
+| Host A | 90 | 1024 MB | 10 * 90 + 25 = 925 | 10 * 100 + 25 = 1025 | 10 * 2 + 0 = 20 |
+| Host B | 50 | 2048 MB | 10 * 50 + 50 = 550 | 10 * 55 + 50 = 600 | 10 * 1 + 1 = 11 |
+| Host C | 10 | 4096 MB | 10 * 10 + 100 = 200 | 10 * 11 + 100 = 210 | 10 * 0 + 2 = 2 |
+
+
+### Benefit to oVirt
+
+The multiplier factor in the Scheduling policy edit dialog will become predictable. This will also allow the user to fine tune his preferences (for example: cpu vs. memory vs. weak affinity rules).
+
+### Implementation details
+
+#### Scheduling process changes
+
+The existing policy units won't be touched. The normalization process will be done by adding an extra step to the current Filter / Weight sequence. It will become Filter / Weight / Select.
+
+#### User Experience
+
+No changes to the UI or REST will be necessary, this change will be totally transparent.
+
+#### Installation/Upgrade
+
+Users using the weight multipliers might need to update them to reasonable values.
+
+### Dependencies / Related Features
+
+This normalization is needed as a ground work for all weak affinity features, including host affinity and Rack based balancing.
+
+### Documentation / External references
+
+- [Bug: Soft negative affinity](https://bugzilla.redhat.com/show_bug.cgi?id=1207255)
+- [Bug: Normalize policy unit weights](https://bugzilla.redhat.com/show_bug.cgi?id=1306263)
+
+### Testing
```

Comment:
The following was added to the testing sections:
"The selection step result is visible in the DEBUG log (logger name: org.ovirt.engine.core.bll.scheduling.policyunits.RankSelectorPolicyUnit) in a CSV table form" etc.